systems/ps3.ixx

```diff
@@ -0,0 +1,262 @@
+			uint16_t next_key;
+			uint16_t key_len;
+			uint8_t value_fmt;
+			uint32_t value_len;
+			uint32_t value_pos;
+			for(int i = 0; i < param_count; ++i)
+			{
+				key_pos = static_cast<uint16_t>(data[0x14 + i * 0x10]) |
+						  (static_cast<uint16_t>(data[0x15 + i * 0x10]) << 8);
+				value_fmt = static_cast<uint8_t>(data[0x17 + i * 0x10]);
+				value_len = static_cast<uint32_t>(data[0x18 + i * 0x10]) |
+							(static_cast<uint32_t>(data[0x19 + i * 0x10]) << 8) |
+							(static_cast<uint32_t>(data[0x1A + i * 0x10]) << 16) |
+							(static_cast<uint32_t>(data[0x1B + i * 0x10]) << 24);
+				value_pos = static_cast<uint32_t>(data[0x20 + i * 0x10]) |
+							(static_cast<uint32_t>(data[0x21 + i * 0x10]) << 8) |
+							(static_cast<uint32_t>(data[0x22 + i * 0x10]) << 16) |
+							(static_cast<uint32_t>(data[0x23 + i * 0x10]) << 24);
+				next_key = static_cast<uint16_t>(data[0x14 + (i + 1) * 0x10]) |
+						   (static_cast<uint16_t>(data[0x15 + (i + 1) * 0x10]) << 8);
+				
+				if(i == param_count - 1)
+					key_len = value_table - key_table - key_pos;
+				else
+					key_len = next_key - key_pos;
+
+				if (data.size() < key_table + key_pos + key_len)
+					return sfo;
+				key.assign(reinterpret_cast<char*>(&data[key_table + key_pos]), key_len);
+				
+				if (value_fmt == 0x04)
+				{
+					if (data.size() < value_table + value_pos + 4)
+						return sfo;
+					value_num = static_cast<uint32_t>(data[value_table + value_pos]) |
+								(static_cast<uint32_t>(data[value_table + value_pos + 1]) << 8) |
+								(static_cast<uint32_t>(data[value_table + value_pos + 2]) << 16) |
+								(static_cast<uint32_t>(data[value_table + value_pos + 3]) << 24);
+					value_str.assign(std::to_string(value_num));
+				}
+				else
+				{
+					if (data.size() < value_table + value_pos + value_len)
+						return sfo;
+					value_str.assign(reinterpret_cast<char*>(&data[value_table + value_pos]), value_len);
+				}
+
+				sfo.emplace(trim(erase_all(key, '\0')), trim(erase_all(value_str, '\0')));	
+			}
+		}
+
+		return sfo;
+	}
+
+
+	std::pair<std::string, std::string> deduceSerial(std::string title_id) const
+	{
+		std::pair<std::string, std::string> serial;
+
+		if(title_id.size() == 9)
+		{
+			serial.first = title_id.substr(0, 4);
+			serial.second = title_id.substr(4, 5);
+		}
+		else if(title_id.size() == 10 && title_id[4] == '-')
+		{
+			serial.first = title_id.substr(0, 4);
+			serial.second = title_id.substr(5, 5);
+		}
+
+		return serial;
+	}
+
+
+	std::string deduceFirmware(std::shared_ptr<iso9660::Entry> root_directory, std::string pup_file) const
+	{
+		std::string fw;
+
+		auto pup_entry = root_directory->subEntry(pup_file);
+		if(pup_entry)
+		{
+			auto data = pup_entry->read();
+			if(data.size() < 0x40)
+				return fw;
+
+			uint16_t index = (static_cast<uint16_t>(data[0x3E]) << 8) | data[0x3F];
+			if(data.size() < index + 4)
+				return fw;
+
+			fw.assign(reinterpret_cast<char*>(&data[index]), 4);
+		}
+
+		return fw;
+	}
+};
```

Comment:
This allocates/reads >100Mb chunk just to read byte 0x3E and 0x3F.
I don't think there is any benefit of reading disc firmware update version using dumper, this code can be removed.